roles/jd-server/src/lib/job_declarator/mod.rs

```diff
@@ -280,23 +280,21 @@ impl JobDeclaratorDownstream {
                                         match Self::collect_txs_in_job(self_mutex.clone()) {
                                             Ok(_) => {
                                                 info!("All transactions in downstream job are recognized correctly by the JD Server");
-                                                let hexdata =
-                                                    match JobDeclaratorDownstream::get_block_hex(
-                                                        self_mutex.clone(),
-                                                        message,
-                                                    ) {
-                                                        Ok(inner) => inner,
-                                                        Err(e) => {
-                                                            error!(
+                                                match JobDeclaratorDownstream::get_block_hex(
```

Comment:
Inside `get_block_hex` we call `collect_txs_in_job`(again, after the call in L280) which uses the mutex heavily, that could be something to improve in a future commit

Reply:
hmm I see your point but I don't think there's an easy way out of this

in theory we could do something like this:
```diff
    fn get_block_hex(
        self_mutex: Arc<Mutex<Self>>,
+       transactions_list: Vec<bitcoin::Transaction>
        message: SubmitSolutionJd,
    ) -> Result<String, Box<JdsError>> {
        let (last_declare_, _, _) = self_mutex
            .clone()
            .safe_lock(|x| x.declared_mining_job.clone())
            .map_err(|e| Box::new(JdsError::PoisonLock(e.to_string())))?;
        let last_declare = last_declare_.ok_or(Box::new(JdsError::NoLastDeclaredJob))?;
-         let transactions_list = Self::collect_txs_in_job(self_mutex)?;
        let block: Block =
            roles_logic_sv2::utils::BlockCreator::new(last_declare, transactions_list, message)
                .into();
        Ok(hex::encode(serialize(&block)))
    }
```

and then just take advantage of the fact that this call to `get_block_hex` happens on this context:
```diff
                                        match Self::collect_txs_in_job(self_mutex.clone()) {
-                                             Ok(_) => {
+                                             Ok(transactions_list) => {
                                                info!("All transactions in downstream job are recognized correctly by the JD Server");
                                                match JobDeclaratorDownstream::get_block_hex(
                                                    self_mutex.clone(),
+                                                   transactions_list
                                                    message,
                                                ) {
```

the problem is that there's yet another call to `get_block_hex` here:
https://github.com/stratum-mining/stratum/blob/5d569bbdb0f6d1dc7bc97a62abff9d6ded4b425d/roles/jd-server/src/lib/job_declarator/mod.rs#L319

and if we applied the changes above to `get_block_hex`, then we would need an extra call to `collect_txs_in_job` here, which brings us back to the original point

Reply:
I do agree this is meaningful technical debt, but I'm afraid that mitigating it on this PR would snowball into a massive change that would probably fall out of scope

unless we can find a simple fix, it should become something to be addressed when we move onto refactoring the application layer (which should only be our focus after we're done with the low-level/protocols layer)

Reply:
Yea, I agree that it should be handled in a different PR